### ISSUE TITLE
Generations taken from Train Set (CUB200 dataset)

### ISSUE THREAD

**Issue by SerezD** [lucidrains/DALLE-pytorch#355]
Hi everyone, thank you for the amazing job on this repository. 
I'm trying to replicate @kobiso work on CUB200 dataset (#131)  on a 2 GPU architecture. I'm not directly using your script "train_dalle.py" but a very similar one. The main difference is that I'm using data parallel to train DALLE in a multi-gpu environment.

Here are some details of my implementation:

```
# Step 1: load pretrained VQGAN. SET PARAMETERS
vae = VQGanVAE()

# Define Parameters
DIM = 256
TEXT_SEQ_LEN = 80
DEPTH = 8
HEADS = 8
DIM_HEAD = 64
ATTN_TYPES = ('full', 'axial_row', 'axial_col', 'conv_like')

# Training Parameters
LEARNING_RATE = 0.0006
EPOCHS = 500                      # but model automatically stops after 300 epochs
BATCH_SIZE = 64          

# Adamw Optimizer. 
BETAS = (0.9, 0.96)
EPS = 1e-08
WEIGHT_DECAY = 4.5e-2

# gradient clipping
MAX_NORM = 0.5  

# Scheduler
FACTOR = 0.5               
PATIENCE = 5                
COOLDOWN = 10               
MIN_LEARNING_RATE = LEARNING_RATE * (FACTOR**6)   # allow lr reduction 6 times

# Step 2: Dataset and tokenizer
IMAGE_SIZE = vae.image_size

ds = TextImageDataset(
        folder=dataset_path,
        text_len=TEXT_SEQ_LEN,
        image_size=IMAGE_SIZE,
        truncate_captions=True,
        tokenizer=tokenizer,        # default tokenizer
        shuffle=True,
    )

dl = DataLoader(ds, batch_size=BATCH_SIZE, shuffle=True, drop_last=True)

# Step 3: define DALLE
dalle = DALLE(
    vae=vae,  # pretrained
    num_text_tokens=tokenizer.vocab_size,
    text_seq_len=TEXT_SEQ_LEN,
    dim=DIM,
    depth=DEPTH,
    heads=HEADS,
    dim_head=DIM_HEAD,
    attn_types=ATTN_TYPES
).cuda()

# data parallel on 2 gpu
dalle = torch.nn.DataParallel(dalle, device_ids=gpu_cores)


# optimizer:
def group_weight(model):
    group_decay, group_no_decay = [], []
    for params in model.named_parameters():
        if 'transformer' in params[0]:
            if 'bias' in params[0] or 'norm' in params[0]:
                group_no_decay.append(params[1])
                continue
        group_decay.append(params[1])

    assert len(list(model.parameters())) == len(group_decay) + len(group_no_decay)
    groups = [dict(params=group_decay), dict(params=group_no_decay, weight_decay=.0)]
    return groups

opt = torch.optim.AdamW(group_weight(dalle), lr=LEARNING_RATE, betas=BETAS, eps=EPS, weight_decay=WEIGHT_DECAY)

# ReduceLRonPlateau
scheduler = ReduceLROnPlateau(opt, mode='min', factor=FACTOR, patience=PATIENCE, cooldown=COOLDOWN,
                                                     min_lr=MIN_LEARNING_RATE, verbose=True)

```
This is the loss-plot obtained during training. Learning rate reductions are clearly visible.

![W B Chart 27_8_2021, 11_30_03](https://user-images.githubusercontent.com/68145713/131105985-379d2e9f-c586-479e-9e6d-65e02285b045.png)

Briefly speaking, my problem is:
When I try to generate images from new captions, in some cases everything looks fine (in the following image I'm showing best 3/64 according to CLIP - the generations aren't as decent as @kobiso 's, but acceptable):

![0](https://user-images.githubusercontent.com/68145713/131106681-0b368a2d-6029-41d8-99d7-1c1bd37d842f.png)

In other cases, the images are direct reproductions of training set images! For example, the first image of the following generation is taken from file Clay_Colored_Sparrow_0104_110699.jpg
![24](https://user-images.githubusercontent.com/68145713/131106837-66d00739-a582-4c77-b760-845e3a1ccfb8.png)

Clay_Colored_Sparrow_0104_110699.jpg :
![Clay_Colored_Sparrow_0104_110699](https://user-images.githubusercontent.com/68145713/131108052-81ed339d-d8ee-4565-8d9f-7f9931a25f53.jpg)

Where some captions of this particular image are quite similar to the one I've passed as an input:
> the bird has a small bill that is orange and black.
> the wings are tri-colored, grey, russet, and brown, with similar markings on cheek patch and eyebrow.
> a small bird with a white throat and belly, with brown, black and white feathers covering its wings.
> this bird has a white belly and breast, with a brown superciliary and sharp bill.
> a small bird with a white belly and brown head and wings.
> a small bird with brown, black, and white speckled wings and orange beak and feet.
> this bird has wings that are brown and black with a white belly
> this colorful bird has an orange bill, orange feet, and brown and orange wings.
> this bird has a light brown and dark grey striped crown, an orange beak, and a light grey breast.
> this bird has grey coloring on its breast and belly, and brown and black strips on its wings.

Maybe someone can help me understand if this is a normal behaviour or not, and how to debug this ? 

Thank you.

**Update**
I have made some tests with basic captions. The behaviour still seems to appear. My idea is that training Dall-E on a small dataset like CUB doesn't give the model a great generalization ability. 
@lucidrains @kobiso  can you confirm that ? Here I post some examples

![0](https://user-images.githubusercontent.com/68145713/131710519-927d5f63-11d9-4c2e-bca6-de550fd53990.png)

![11](https://user-images.githubusercontent.com/68145713/131710562-01518e27-4958-4a85-a4b2-8c0607c97bb6.png)

![23](https://user-images.githubusercontent.com/68145713/131710652-854ca6c9-bc57-4ede-8f7e-d7f5624c1cc1.png)

![38](https://user-images.githubusercontent.com/68145713/131710773-cf1f482d-1188-4bd4-bcbc-e52c7d0e5e98.png)
